x-pack/filebeat/input/mito/config_auth.go

```diff
@@ -0,0 +1,297 @@
+		if err != nil {
+			return nil, fmt.Errorf("oauth2 client: error loading credentials: %w", err)
+		}
+		return oauth2.NewClient(ctx, creds.TokenSource), nil
+	default:
+		return nil, errors.New("oauth2 client: unknown provider")
+	}
+}
+
+// getTokenURL returns the TokenURL.
+func (o *oAuth2Config) getTokenURL() string {
+	switch o.getProvider() {
+	case oAuth2ProviderAzure:
+		if o.TokenURL == "" {
+			return endpoints.AzureAD(o.AzureTenantID).TokenURL
+		}
+	}
+
+	return o.TokenURL
+}
+
+// getProvider returns provider in its canonical form.
+func (o oAuth2Config) getProvider() oAuth2Provider {
+	return o.Provider.canonical()
+}
+
+// getEndpointParams returns endpoint params with any provider ones combined.
+func (o oAuth2Config) getEndpointParams() url.Values {
+	switch o.getProvider() {
+	case oAuth2ProviderAzure:
+		if o.AzureResource != "" {
+			if o.EndpointParams == nil {
+				o.EndpointParams = url.Values{}
+			}
+			o.EndpointParams["resource"] = []string{o.AzureResource}
+		}
+	}
+
+	return o.EndpointParams
+}
+
+// Validate checks if oauth2 config is valid.
+func (o *oAuth2Config) Validate() error {
+	if !o.isEnabled() {
+		return nil
+	}
+
+	switch o.getProvider() {
+	case oAuth2ProviderAzure:
+		return o.validateAzureProvider()
+	case oAuth2ProviderGoogle:
+		return o.validateGoogleProvider()
+	case oAuth2ProviderDefault:
+		if o.TokenURL == "" || o.ClientID == "" || o.ClientSecret == "" {
+			return errors.New("both token_url and client credentials must be provided")
+		}
+		if (o.User != "" && o.Password == "") || (o.User == "" && o.Password != "") {
+			return errors.New("both user and password credentials must be provided")
+		}
+	default:
+		return fmt.Errorf("unknown provider %q", o.getProvider())
+	}
+
+	return nil
+}
+
+// findDefaultGoogleCredentials will default to google.FindDefaultCredentials and will only be changed for testing purposes
+var findDefaultGoogleCredentials = google.FindDefaultCredentials
+
+func (o *oAuth2Config) validateGoogleProvider() error {
+	if o.TokenURL != "" || o.ClientID != "" || o.ClientSecret != "" ||
+		o.AzureTenantID != "" || o.AzureResource != "" || len(o.EndpointParams) > 0 {
+		return errors.New("none of token_url and client credentials can be used, use google.credentials_file, google.jwt_file, google.credentials_json or ADC instead")
+	}
+
+	// credentials_json
+	if len(o.GoogleCredentialsJSON) > 0 {
+		if o.GoogleDelegatedAccount != "" {
+			return errors.New("google.delegated_account can only be provided with a jwt_file")
+		}
+		return nil
+	}
+
+	// credentials_file
+	if o.GoogleCredentialsFile != "" {
+		if o.GoogleDelegatedAccount != "" {
+			return errors.New("google.delegated_account can only be provided with a jwt_file")
+		}
+		return o.populateCredentialsJSONFromFile(o.GoogleCredentialsFile)
+	}
+
+	// jwt_file
+	if o.GoogleJWTFile != "" {
+		return o.populateCredentialsJSONFromFile(o.GoogleJWTFile)
```

Comment:
It would be nice to support reading the value directly from config as related to https://github.com/elastic/integrations/issues/2834.

Reply:
This is done now.